controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,467 @@
+}
+
+type SnapshotMonitor struct {
+	*baseMonitor
+
+	nodeName      string
+	eventRecorder record.EventRecorder
+
+	snapshotChangeEventChan <-chan *SnapshotChangeEvent
+	snapshotVerifyTaskChan  chan *snapshotVerifyTask
+
+	inProgressSnapshotVerifyTasks     map[string]struct{}
+	inProgressSnapshotVerifyTasksLock sync.RWMutex
+
+	settings map[string]interface{}
+
+	workers int
+
+	syncCallback func(key string)
+
+	proxyConnCounter       util.Counter
+	lastSnapshotVerifiedAt metav1.Time
+}
+
+func NewSnapshotMonitor(logger logrus.FieldLogger, ds *datastore.DataStore, nodeName string, eventRecorder record.EventRecorder,
+	snapshotChangeEventChan <-chan *SnapshotChangeEvent, setting map[string]interface{}, syncCallback func(key string)) (*SnapshotMonitor, error) {
+
+	ctx, quit := context.WithCancel(context.Background())
+
+	m := &SnapshotMonitor{
+		baseMonitor: newBaseMonitor(ctx, quit, logger, ds, 0),
+
+		nodeName:      nodeName,
+		eventRecorder: eventRecorder,
+
+		snapshotChangeEventChan: snapshotChangeEventChan,
+		snapshotVerifyTaskChan:  make(chan *snapshotVerifyTask, 1),
+
+		inProgressSnapshotVerifyTasks: map[string]struct{}{},
+
+		settings: setting,
+
+		workers: snapshotVerifyWorkerMax,
+
+		syncCallback:     syncCallback,
+		proxyConnCounter: util.NewAtomicCounter(),
+	}
+
+	go m.Start()
+
+	return m, nil
+}
+
+func (m *SnapshotMonitor) Start() {
+	for i := 0; i < m.workers; i++ {
+		go m.snapshotVerifyWorker()
+	}
+
+	go m.processSnapshotChangeEvent()
+	go m.processPeriodicSnapshotVerify()
+}
+
+func (m *SnapshotMonitor) processSnapshotChangeEvent() {
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.logger.WithField("monitor", monitorName).Info("Closing periodic snapshot verification process")
+			return
+		default:
+			event := <-m.snapshotChangeEventChan
+
+			if event.SnapshotName == "" {
+				engine, err := m.ds.GetVolumeCurrentEngine(event.VolumeName)
+				if err != nil {
+					m.logger.WithField("monitor", monitorName).WithError(err).Errorf("failed to find engine for volume %v", event.VolumeName)
+					continue
+				}
+				m.populateEngineSnapshots(engine)
+			} else {
+				m.snapshotVerifyTaskChan <- &snapshotVerifyTask{
+					volumeName:   event.VolumeName,
+					snapshotName: event.SnapshotName,
+				}
+			}
+		}
+	}
+}
+
+func (m *SnapshotMonitor) processPeriodicSnapshotVerify() {
+	ticker := time.NewTicker(snapshotVerifyProcessPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.logger.WithField("monitor", monitorName).Info("Closing periodic snapshot verification process")
+			return
+		case <-ticker.C:
+			v := m.settings[string(types.SettingNameSnapshotDataIntegrity)]
+			dataIntegrity := v.(string)
+
+			v = m.settings[string(types.SettingNameSnapshotDataIntegrityInterval)]
+			dataIntegrityInterval := v.(int64)
+
+			if dataIntegrity == string(types.SnapshotDataIntegrityDisabled) ||
+				!time.Now().After(m.lastSnapshotVerifiedAt.Time.Add(time.Duration(dataIntegrityInterval)*time.Minute)) {
```

Comment:
It seems that some newly created snapshots should wait for `dataIntegrityInterval` before starting the hash.
The pro is the resync/re-verification mechanism is straightforward. But the con is that the calculation for all snapshots is started at one single moment hence there will be relatively high CPU & Disk usage since that moment.

Reply:
For the newly created snapshots, they are sents to workers for hashing and don't need to wait for `dataIntegrityInterval`. The path is 

`Node controller (aware of snapshot update) -> enqueueSnapshot -> snapshotChangeEventChan -> snapshotVerifyTaskChan -> workers`

Reply:
OK. Got it. The newly created snapshots should be fine. But the pro and con I mentioned still exist.

Reply:
Yeah, you're right. I created a [ticket](https://github.com/longhorn/longhorn/issues/4765) to throttle the background io for reducing the performance impact. I think this will be helpful for the con.